### ISSUE TITLE
Optimize app retrieval by ID 

### ISSUE THREAD

**Issue by andymarch** [okta/terraform-provider-okta#93]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

data_source_okta_app uses the /apps end point with a filter to identify apps regardless of the information provided in the filter. The rate limit for create and list applications is significantly lower than most other endpoints (25 for most orgs, going up to 100 with enterprise license). There is a separate rate limit for /apps/{id} which falls into a different and much larger rate limit bucket. 

This is specifically called out as the example on the[ rate limits page](https://developer.okta.com/docs/reference/rate-limits/).

> For example, the limit for getting an application by ID is higher than the more general limit for the /api/v1/apps endpoint.

If the provided filter contains an id then the more specific /apps/{id} should be used rather than using the filtered list.

Happy to take a run at this myself, it looks like the best place for a change would be in [app_filter](https://github.com/terraform-providers/terraform-provider-okta/blob/9ae5cc400dab482fc4d43a798e1295b00712185c/okta/app_filter.go).

### New or Affected Resource(s)

* okta_app

### Potential Terraform Configuration

```hcl
# Should list all applications with a filter on the label
data "okta_app" "anapp" {
  label = "An App"
}

# Currently this performs a list with an id filter
data "okta_app" "anotherapp" {
  id = "0oa9k05yf3gGKOEV30x6"
}
```

### References

* #35 - Indicates similar issues with managing app users.

**Comment by noinarisak** [comment]
@andymarch Could you let us know if PR #97 resolved this issue. Thanks!

**Comment by adw1n** [comment]
@noinarisak it did not. I'm using the latest v3.4.1 relase and I see a ton of requests being made to /api/v1/apps which consitently causes rate limit violations for my Org. This furthermore causes downtime for my app which relies on being able to use `/api/v1/apps` endpoint. I had a look at https://github.com/oktadeveloper/terraform-provider-okta/pull/97 description and

> Improve app filtering by using ?q=<label> rather than listing and filtering. App doesn't have good filtering like users do so this was the best approach.

apporach sounds really problemtatic. The way I see it there at least three approaches:
1. you can list all apps once and then iterate over the results in a for loop and compare app label / id. You'd be making `TOTAL_NUMBER_OF_APPS / MAX_PAGINATION_LIMIT = TOTAL_NUMBER_OF_APPS / 200` requests to Okta.
2. make a `/api/v1/apps?q=<label>` request to Okta for every app (?)
3. use `/apps/{id}` like this ticket suggests

`#3` is clearly a winner

Please have this ticket reopened.

**Comment by adw1n** [comment]
On second thought I don't even use `data_source_okta_app` so it's something else in the provider code DDOSing `/api/v1/apps` endpoint.

**Comment by adw1n** [comment]
I dig deeper using `TF_LOG` and it turns out these three endpoints are giving me a hard time:
1. GET /api/v1/apps/{app_id}/groups?limit=200
2. GET /api/v1/apps/{app_id}/users?limit=200
3. GET /api/v1/apps/{app_id}/groups/{group_id}

They all share the same `/api/v1/apps` very low rate limit.

**Comment by bogdanprodan-okta** [comment]
This will be fixed in the upcoming v3.8.0.
